DietPi-Software | Rework dependency and choice system

nonnorm opened the pull request:
<!--
Before submitting a pull request:
- Please ensure the target branch is "dev" (active development): https://github.com/MichaIng/DietPi/tree/dev
- Please ensure changes have been tested and verified functional.
-->
**Status**: Testing
- [x] Migrate all software to use new system
- [x] ~Add dummy IDs for webserver/desktop/browser~ **EDIT: Not used anymore. Another way to resolve such meta dependencies has been implemented.**
- [x] Add own Firefox install option
- [x] Add browser meta dependency, used by desktops
- [x] Add desktop default menu
- [x] Add browser default menu

**Reference**: https://github.com/MichaIng/DietPi/pull/3921#issuecomment-730694783

**Commit list/description**:
<!--
- DietPi-Config | Add "Fan control" option to "Performance Options"
![Screenshot](https://xxx.github.com/images/xxx.png)
- DietPi-Config | Add "Fan control" support for Odroid C2
- DietPi-Config | Syntax fix
-->
- DietPi-Software | Begin to rework dependency system
- DietPi-Software | Begin to switch to new system

nonnorm committed:
v7.2
Init v7.2

nonnorm committed:
v7.2
+ DietPi-Software | Begin to rework dependency system

nonnorm committed:
v7.2
+ DietPi-Software | Fix loop

nonnorm committed:
v7.2
+ DietPi-Software | Fix array and conditional

nonnorm committed:
v7.2
+ DietPi-Software | Begin to switch to new system

nonnorm committed:
v7.2
+ DietPi-Software | Add missing !

nonnorm committed:
v7.2
+ DietPi-Software | Remove $

MichaIng committed:
v7.2

+ Attempt to resolve merge conflict

MichaIng committed:
v7.2

+ Attempt to resolve merge conflict

MichaIng committed:
v7.2 (#4316)

+ META | Update "dependencies" branch

MichaIng commented:
Awesome start with this, many thanks!

- We can corporate all this "Software that requires XYZ"/"Software required by XYZ" stuff as well.
- For webserver and desktop, I through about some kind of "OR" operator. But in case of webserver, it needs to be handled specially anyway to respect the preference system. And I want a desktop preference menu as well, as currently LXDE is always chosen, if no desktop is installed or selected but required. _Same could follow for browsers, after making Firefox an own install option._ We could add a dummy webserver and desktop software ID. In the loop we then check the these IDs specifically to handle them, e.g. in a different function to keep it clean, which basically changes the `$j` ID based on the preference. An important difference for those is that webservers and desktops are never reinstalled (`< 1`), while all other dependencies are (`!= 1`). We should keep that behaviour for now until we're able to rework them to not overwrite existing configs.
- And we need to do the loop incrementally. My idea was to make inner dependency loop `for j in "${aSOFTWARE_REQUIRES[@]}"` an own function with the current outer loop software ID as input, which `continue`s, if a dependency is already set to `1`, else calls itself with the current dependency as input. The `== 1 )) && continue` is important then to avoid unlimited nesting.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -304,8 +304,7 @@ _EOF_
 		aSOFTWARE_ONLINEDOC_URL[$software_id]='https://dietpi.com/docs/software/desktop/#lxde'
-		aSOFTWARE_REQUIRES_ALSA[$software_id]=1
-		aSOFTWARE_REQUIRES_XSERVERXORG[$software_id]=1
+		aSOFTWARE_REQUIRES=(5 6)
```
We need to make it an array: `aSOFTWARE_REQUIRES[$software_id]`
Since nested arrays are not possible in bash, the dependencies need to be a simple string:
```
aSOFTWARE_REQUIRES[$software_id]='5 6'
```

The loop hence needs to be:
```
for j in ${aSOFTWARE_REQUIRES[$i]}
```
without quotes. Shellcheck will complain, but we must ignore it (`# shellcheck disable=SCXXXX`). At least currently I have no better idea how to the each dependency ID an array value without making it much more complicated.

nonnorm commented:
> We could add a dummy webserver and desktop software ID.

What about automated upgrades and user input?

nonnorm committed:
v7.1
+ DietPi-Software | Switch built-in dependencies to new system

nonnorm committed:
v7.2
+ DietPi-Software | Update showings in list

nonnorm committed:
v7.2
+ DietPi-Software | Fix spacing

nonnorm committed:
v7.2
+ DietPi-Software | Rename git

nonnorm committed:
v7.2
+ DietPi-Software | Add Resolve_Dependencies function

nonnorm committed:
v7.2
+ DietPi-Software | Fix loop

nonnorm committed:
v7.2
+ DietPi-Software | Make all software use new dependency system

nonnorm committed:
Merge branch 'dev' into dependencies

nonnorm committed:
v7.2
+ DietPi-Software | Add dependencies for webserver stacks

nonnorm committed:
Merge branch 'dependencies' of https://github.com/MichaIng/DietPi into dependencies

nonnorm committed:
v7.2
+ DietPi-Software | CodeFactor fixes

nonnorm committed:
v7.2
+ DietPi-Software | Add Firefox and Lynx web browsers as own install options

nonnorm committed:
v7.2
+ DietPi-Software | Add dummy software IDs

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -285,6 +285,7 @@ _EOF_
 			'●─ Development & Programming ' #26
 			'●─ Text Editors ' #27
 			'●─ Desktop Utilities ' #28
+			'●- Browsers ' #29
```
Please add browsers below "Desktop Utilities" for now. We cause a lot required changes in the docs, including webserver redirects to not break visitors links, when changing categories, so I want to do category changes in dedicated PRs, when required. In the docs, other than here, Chromium is already below "Desktop Utilities", so for that no change is required 😉.

Maybe your plan about how to implement the browser preference implies the category? I think for now we can do it without a dedicated category, based in software ID only, as well. Also Lynx is not a browser for a desktop, so it does not fit well into a browser preference that is followed when selecting a desktop IMO 😉.

nonnorm replied:
I implemented Lynx because I thought that it would be best to have another software for the browser preference.  If we keep it, which category would we put it in?

MichaIng replied:
Generally it's good to have some more choices, but not a requirement to allow the choice, even when it's only between two options. Since currently I have no further idea than using the browser preference for desktops, all options should be reasonable for desktop usage, where I'm not sure if anyone would use a console-based browser? Although it would be very lightweight, I guess, probably I should try it out :).

Lynx could fit into network tools? But, as long as we indeed do not add it for desktop browser choice, probably it's not worth to keep it, as long as we don't do anything reasonable more than installing the APT package, what you think? At least I was thinking to remove most of the other network tools, probably after adding some documentation troubleshooting/monitoring section where they are used for example. Everyone who knows how to use those tools, definitely knows how to install them the native way via APT.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -689,24 +679,24 @@ _EOF_
 		aSOFTWARE_ONLINEDOC_URL[$software_id]='https://dietpi.com/docs/software/media/#komga'
-		aSOFTWARE_REQUIRES_JAVA_JRE_JDK[$software_id]=1
+		aSOFTWARE_REQUIRES[$software_id]='8'
+		a
```
Typo.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -1959,6 +1941,31 @@ _EOF_
+		#------------------
+		software_id=190
+
+		aSOFTWARE_NAME[$software_id]='Firefox'
```
Thanks for that 👍, makes sense to have this as a separate option in any case. It should be even possible to install and run it in a dedicated X server session, without a wrapping desktop.

MichaIng commented:
> What about automated upgrades and user input?

At least I'm not sure how to reasonably integrate them into the software ID based dependency loop, so leave it as it is for now? But if you have an idea how to make it better, I'm open for that. Probably we find a more consistent looking way to deal with such software-specific needs and can include RAM usage (in general or for compilation => increase swap file when required) and GPU memory (strictly required + recommended, the first being applied automatically like currently for Kodi, the second as interactive dialogue after installs) and when an autostart option is available (to offer changing it).

MichaIng committed:
v7.2

+ DietPi-Software | Typo

MichaIng commented:
Just found: https://github.com/MichaIng/DietPi/issues/3683#issuecomment-665037171
Epiphany browser should be an alternative for desktops. It was the default browser in older Raspbian images.

MichaIng committed:
Merge branch 'dev' into dependencies

MichaIng committed:
v7.2

+ DietPi-Software | Assure that aSOFTWARE_REQUIRES assignments do not overwrite each other. We don't need to worry about leading spaces, as the way the string is looped through ignores all spaces anyway.

MichaIng commented:
Or brave browser: https://brave.com/linux/
But it's arm64 only 😢.

WIP for ARMv8: https://github.com/brave/brave-browser/issues/11836

MichaIng committed:
v7.2

+ DietPi-Software | Remove browser category, to not cause too much docs changes for now. In the docs, Chromium was a "Desktop Utility", reasonably, and all browsers fit into this category very well. 
+ DietPi-Software | Remove dummy software IDs for now, as meta dependencies can be resolved without them
+ DietPi-Software | Move GIMP, Xfce Power Manager and browsers into "Desktop Utilities" category, to match docs, minimise overhead and as it simply fits well.
+ DietPi-Software | Remove desktop dependency from VSCodium and QuiteRSS. It requires testing, but it should be possible to start both directly from console via xinit.
+ DietPi-Software | Resolve desktop and webserver dependencies within the new dependency system. Make them meta IDs which are resolved to actual software IDs, based on  install state and user preference, respectively the default. The resulting ID is then handled like every other dependency.
+ DietPi-Software | Ask for automated APT upgrades only once. Use a flag variable which contains the last software ID which recommends automated upgrades and ask user once after all dependencies have been resolved.
+ DietPi-Software | Make the $j loop variable in the Resolve_Dependencies() function a local one, so that recursive calls do not overwrite it.
+ DietPi-Software | Remove Lynx for now. It's a console browser and hence not a reasonable browser preference (ToDo) choice, which aims to be used for desktop installs.

MichaIng commented:
@ravenclaw900 
I added a large commit. Please have a look if you agree. While merging webserver and desktop meta dependencies into the new system, the dummy software titles were actually not required, so I removed them. Current implementation is to add "webserver" or "desktop" (and ToDo "browser") as word to the `aSOFTWRE_REQUIRES[<ID]` string and within the recursive dependency loop it is replaced with an actual software ID, based on install state and preference choice (respectively default, in case of desktops currently). This way all recursive dependencies are resolved the same way, and the info output is the same as well "xyz will be installed".

As discussed, to avoid too much required changes in the docs, I moved browsers into the "Desktop Utilities" category, together with all other desktop/GUI applications. And I removed Lynx for now. At least on x86_64 we could also add Brave as a real desktop browser: https://brave.com/linux/#linux

MichaIng committed:
v7.2

+ DietPi-Software | Syntax

nonnorm commented:
Sorry, I've had little time to work on this recently.  Anyway, it looks good.  Yes, let's add Brave for `x86_64`.  Thinking for the future, should we replace "Chromium autostart" with "Browser autostart", or maybe even "Desktop app autostart" and let people select from installed things in the "Desktop Utilities" category?

MichaIng commented:
No problem. I'll go on with the browser meta dependency and the related choice menus.

MichaIng committed:
v7.2

+ DietPi-Software | Add browser meta dependency for desktops, default to Firefox for now, preference choice TBD
+ DietPi-Software | Reduce install state file writes by collecting content in a variable first and write everything in a single chunck
+ DietPi-Software | Move Resolve_Dependencies() function up to Mark_Dependencies() function
+ DietPi-Software | Use database array software names and descriptions for choice and preference menus

MichaIng committed:
v7.2

+ DietPi-Survey_report | Add support for Firefox with v7.2 and remove DietPi-NordVPN, forgotten with v7.1

MichaIng committed:
v7.2

+ README | Add Firefox source code to 3rd party links

MichaIng committed:
v7.2

+ DietPi-Software | Start implementing desktop and browser preference system
+ DietPi-Software | WiFi/Tor HotSpot: When HotSpot criteria are not fulfilled, only unmark "to be installed" mark (1) but do not mark "already installed" mark, e.g. when a hotspot has been set but the adapter temporarily unplugged or connection issue.
+ DietPi-Software | Use aSOFTWARE_NAME[<ID>] variable where possible instead of hardcoded name in menus
+ DietPi-Software | Minor coding

MichaIng committed:
v7.1

+ DietPi-Software | Apply desktop and browser preference via dependency system
+ DietPi-Software | Explain preference systems in initial selection menu instead of in another separate prompt
+ DietPi-Software | Fix Nginx install detection when selecting a webserver preference: We install nginx-light and there would be nginx-full and nginx-extra as well, all providing the "nginx" meta package. Check for nginx-common, which is pulled by all of them.
+ DietPi-Software | Fix docs links from help prompt
+ DietPi-Software | Minor coding

MichaIng committed:
v7.2

+ DietPi-Software | Syntax

MichaIng committed:
v7.2

+ DietPi-Software | Syntax

MichaIng committed:
v7.2

+ DietPi-Software | Fix disabling dietpi-desktop_setup: dash's echo does not support the "-e" option but expands escape sequences automatically

MichaIng committed:
v7.2

+ DietPi-Patches | Fix disabling dietpi-desktop_setup

MichaIng committed:
v7.2

+ dietpi.txt | Add desktop and browser preference settings and enhance explanation of such preferences which are only installed when another software requires it

MichaIng committed:
v7.2

+ DietPi-Software | Respect new dietpi.txt settings for desktop and browser preferences and apply intended default when the webserver preference setting is missing
+ DietPi-Software | Apply preference indices when related software title is installed to avoid the situation that e.g. Xfce is manually installed but LXDE still shown as desktop preference.

MichaIng committed:
v7.2

+ DietPi-Software | Enhance final install prompt a bit

MichaIng committed:
v7.2

+ DietPi-Software | Typo

MichaIng committed:
v7.2

+ DietPi-Software | Do not print software list in lower case and somehow formatted letters anymore, there is no real reason for this
+ DietPi-Software | Rename "X Server" back to "X.Org X Server" as this does not imply issues with CLI list print anymore. Leave "Git" renamed from "Git Client", as there is not really a separation. The "git" package and command can act as both, respectively a local Git repository can be accessed like a server as well.

MichaIng committed:
v7.2

+ DietPi-Software | Try to make pretty clear that webserver/desktop/browser preferences will "only" be installed when another selection depends on it.

MichaIng committed:
v7.2

+ CHANGELOG | DietPi-Software: A desktop preference choice has been added. Before first boot via new dietpi.txt AUTO_SETUP_DESKTOP_INDEX setting or with the new menu entry it can be now chosen which desktop to install when another software selection requires one, especially remote desktop solutions. It works like the webserver preference, so it only play a role when a desktop is installed as dependency and is overridden by manual desktop selections.
+ CHANGELOG | DietPi-Software: A browser preference choice has been added. Before first boot via new dietpi.txt AUTO_SETUP_BROWSER_INDEX setting or with the new menu entry it can be now chosen whether and which browser to install automatically together with desktops. Previously Firefox was installed with every desktop. It only plays a role when a desktop is installed and is overridden by manual browser selections. With this change, Firefox has become an independent software option.
